### PR TITLE
feat: add prettier and solhint for contracts

### LINKED ISSUE #13: add solidity linter + auto formatter

### PR DESCRIPTION
**Pull request by davisshaver**
Adds recommended extensions.

Adds `prettier` command to `contracts` using `prettier-plugin-solidity`.

Also adds `solhint` integration (`pnpm solhint` and `pnpm solhint:check` in the `contracts` repo).

Solhint configuration comes from the [dapptools-template](https://github.com/transmissions11/dapptools-template-1).

Closes #13

## Breakdown of what each extension does

The extension `bradlc.vscode-tailwindcss` adds suggestions and definitions for Tailwind classes

Screen Shot 2022-07-01 at 4 14 11 PM

Screen Shot 2022-07-01 at 4 12 57 PM

The extension `rvest.vs-code-prettier-eslint` lets us format code using the Jest/Prettier mashup config. The extension `dbaeumer.vscode-eslint` is needed to have lint errors show while editing your file.

Screen Shot 2022-07-01 at 6 18 56 PM

Screen Shot 2022-07-01 at 6 18 46 PM

The node package `prettier-plugin-solidity` extend Prettier's functionality for formatting Solidity, in conjunction with the extension `juanfranblanco/vscode-solidity`. [See docs here.](https://github.com/juanfranblanco/vscode-solidity#formatting-using-prettier-and-the-prettier-solidity-plugin) Note: Solhint plugins are not supported yet ([see here](https://github.com/juanfranblanco/vscode-solidity#solhint)), so Prettier errors won't show up as linting issues, although they will be cleaned up as part of formatting.

The extension `redhat.vscode-yaml` shows us formatting issues with YAML files.

Screen Shot 2022-07-01 at 4 25 41 PM

### CODE REVIEW

**Comment by frolic** [comment]
Any advice for getting this to autofix/format in VSCode? Would love to add the proper config settings, recommended plugins, etc. to https://github.com/holic/web3-scaffold/tree/main/.vscode

**Comment by davisshaver** [comment]
@holic I added some recommended extensions but I want to try setting this up in a clean VS Code install to see what the experience is like, I'll share results after I get a chance to do it.

**Comment by jamiew** [comment]
I'll give this a shot shortly! I've had a lot of trouble getting VScode + prettier + solidity to behave consistently

File: .vscode/settings.json [hunk 1]
```diff
@@ -1,5 +1,5 @@
 {
-  "editor.defaultFormatter": "esbenp.prettier-vscode",
+  "editor.defaultFormatter": "rvest.vs-code-prettier-eslint",
```
Comment (frolic) [inline .vscode/settings.json:2]: 🎉 Woo, I think this is the missing piece when working on https://github.com/holic/web3-scaffold/issues/14

File: .vscode/extensions.json [hunk 2]
```diff
@@ -3,7 +3,15 @@
...
   // List of extensions which should be recommended for users of this workspace.
-  "recommendations": ["dbaeumer.vscode-eslint", "esbenp.prettier-vscode"],
+  "recommendations": [
+    "bradlc.vscode-tailwindcss",
+    "dbaeumer.vscode-eslint",
+    "JuanBlanco.solidity",
+    "redhat.vscode-yaml",
+    "rvest.vs-code-prettier-eslint",
+    "tintinweb.solidity-visual-auditor",
+    "tintinweb.vscode-solidity-language"
```
Comment (frolic) [inline .vscode/extensions.json]: What does this one provide? Does the syntax highlighting conflict with `JuanBlanco.solidity`?

File: packages/contracts/package.json [hunk 3]
```diff
@@ -7,14 +7,20 @@
...
+    "solhint:check": "solhint --config ./.solhint.json \"src/**/*.sol\"",
+    "solhint": "pnpm solhint:check --fix"
   },
   "devDependencies": {
     "@ethersproject/abi": "^5.0.0",
     "@ethersproject/bytes": "^5.0.0",
     "@ethersproject/providers": "^5.0.0",
     "@typechain/ethers-v5": "^10.0.0",
     "ethers": "^5.5.2",
+    "prettier-plugin-solidity": "^1.0.0-beta.19",
```
Comment (frolic) [inline packages/contracts/package.json:25]: Any idea if this also runs `solhint --fix` (like prettier + eslint)?

Comment (davisshaver) [reply packages/contracts/package.json:25]: I don't think it does – that would be nice though, I'm going to work some more on this PR and I will investigate how we might run `solhint --fix` in VS Code automatically. I'm going to convert this to a draft PR in the meantime.

File: .vscode/extensions.json (hunk 2 above)
Comment (davisshaver) [reply .vscode/extensions.json]: @holic I don't think it does but I will double check this with a clean VS Code install. I'll also include some documentation/links about what all the extensions do and other recommended VS Code configuration.

File: .solhint.json [hunk 4]
```diff
@@ -0,0 +1,20 @@
+{
+    "extends": "solhint:recommended",
+    "plugins": ["prettier"],
+    "rules": {
+        "code-complexity": ["error", 7],
+        "compiler-version": ["error", "^0.8.9"],
```
Comment (frolic) [inline .solhint.json:6]: I wish this could be auto detected in some way. I feel like we've had to specify this version in a ton of places :(

**Pull request merged**